pkg/apis/options/providers.go

```diff
@@ -0,0 +1,177 @@
+	// ClientSecretFile is the name of the file
+	// containing the OAuth Client Secret, it will be used if ClientSecret is not set.
+	ClientSecretFile string `json:"clientSecretFile,omitempty"`
+
+	// KeycloakConfig holds all configurations for Keycloak provider.
+	KeycloakConfig KeycloakOptions `json:"keycloakConfig,omitempty"`
+	// AzureConfig holds all configurations for Azure provider.
+	AzureConfig AzureOptions `json:"azureConfig,omitempty"`
+	// BitbucketConfig holds all configurations for Bitbucket provider.
+	BitbucketConfig BitbucketOptions `json:"bitbucketConfig,omitempty"`
+	// GitHubConfig holds all configurations for GitHubC provider.
+	GitHubConfig GitHubOptions `json:"githubConfig,omitempty"`
+	// GitLabConfig holds all configurations for GitLab provider.
+	GitLabConfig GitLabOptions `json:"gitlabConfig,omitempty"`
+	// GoogleConfig holds all configurations for Google provider.
+	GoogleConfig GoogleOptions `json:"googleConfig,omitempty"`
+	// OIDCConfig holds all configurations for OIDC provider
+	// or providers utilize OIDC configurations.
+	OIDCConfig OIDCOptions `json:"oidcConfig,omitempty"`
+	// LoginGovConfig holds all configurations for LoginGov provider.
+	LoginGovConfig LoginGovOptions `json:"loginGovConfig,omitempty"`
+
+	// ProviderID should be a unique identifier for the provider.
+	// This value is required for all providers.
+	ProviderID string `json:"providerID,omitempty"`
+	// ProviderType is the OAuth provider
+	// must be set from the supported providers group,
+	// otherwise 'Google' is set as default
+	ProviderType string `json:"provider,omitempty"`
+	// ProviderName is the providers display name
+	// if set, it will be shown to the users in the login page.
+	ProviderName string `json:"providerDisplayName,omitempty"`
+	// ProviderCAFiles is a list of paths to CA certificates that should be used when connecting to the provider.
+	// If not specified, the default Go trust sources are used instead
+	ProviderCAFiles []string `json:"providerCAFiles,omitempty"`
+
+	// LoginURL is the authentication endpoint
+	LoginURL string `json:"loginURL,omitempty"`
+	// RedeemURL is the token redemption endpoint
+	RedeemURL string `json:"redeemURL,omitempty"`
+	// ProfileURL is the profile access endpoint
+	ProfileURL string `json:"profileURL,omitempty"`
+	// ProtectedResource is the resource that is protected (Azure AD only)
+	ProtectedResource string `json:"resource,omitempty"`
+	// ValidateURL is the access token validation endpoint
+	ValidateURL string `json:"validateURL,omitempty"`
+	// Scope is the OAuth scope specification
+	Scope string `json:"scope,omitempty"`
+	// Prompt is OIDC prompt
+	Prompt string `json:"prompt,omitempty"`
+	// ApprovalPrompt is the OAuth approval_prompt
+	// default is set to 'force'
+	ApprovalPrompt string `json:"approvalPrompt,omitempty"`
+	// AllowedGroups is a list of restrict logins to members of this group
+	AllowedGroups []string `json:"allowedGroups,omitempty"`
```

Comment:
Long term I see this moving out of here and into a dedicated AuthZ set of configuration, @NickMeves do you think we should move this out for now or do so later?

Reply:
I'm torn -- because a lot of providers have their own flavors of `AllowedGroups` in addition to this global generic one (e.g. `GitlabProjects`, `GithubRepo`, etc).

While they all somehow populate the `session.Groups` with something that fit's their naming conventions for their provider specific Authorize implementation -- they have different concepts of "Groups" from provider to provider.

For that reason, I lean toward these group settings being per provider I think.

Reply:
leaving inside each provider as Nick suggested.